client-tools/Shell/ProcessMonitoringFeature.cpp

```diff
@@ -49,16 +49,33 @@ void ProcessMonitoringFeature::addMonitorPID(ExternalId const& pid) {
   _monitoredProcesses.push_back(pid);
 }
 
+void ProcessMonitoringFeature::removeMonitorPIDNoLock(ExternalId const& pid) {
+  for (auto it = _monitoredProcesses.begin(); it != _monitoredProcesses.end();
+       ++it) {
+    if (it->_pid == pid._pid) {
+      _monitoredProcesses.erase(it);
+      break;
+    }
+  }
+}
+
+void ProcessMonitoringFeature::moveMonitoringPIDToAttic(
+    ExternalId const& pid, ExternalProcessStatus const& exitStatus) {
+  MUTEX_LOCKER(mutexLocker, _MonitoredExternalProcessesLock);
+  removeMonitorPIDNoLock(pid);
+  _ExitedExternalProcessStatus[pid._pid] = exitStatus;
+}
+
+std::vector<ExternalId>& ProcessMonitoringFeature::getMonitoringVector() {
```

Comment:
This is unsafe. We must return a _copy_ of the vector, not a reference.
If we return a reference, the caller can modify or peek into the vector without the mutex!